asmcomp/selectgen.ml

```diff
@@ -809,19 +894,35 @@ method emit_expr (env:environment) exp =
             env_add_static_exception nfail rs env)
           env handlers
       in
-      let (r_body, s_body) = self#emit_sequence env body in
+      let (r_body, s_body) = self#emit_sequence env body ~bound_name in
       let translate_one_handler (nfail, ids, rs, e2) =
         assert(List.length ids = List.length rs);
+        let ids_and_rs = List.combine ids rs in
         let new_env =
           List.fold_left (fun env ((id, _typ), r) -> env_add id r env)
-            env (List.combine ids rs)
+            env ids_and_rs
+        in
+        let (r, s) =
+          self#emit_sequence new_env e2 ~bound_name:None ~at_start:(fun seq ->
+            List.iter (fun ((var, _typ), r) ->
+                let provenance = VP.provenance var in
+                let var = VP.var var in
+                let naming_op =
+                  Iname_for_debugger {
+                    ident = var;
+                    provenance;
+                    which_parameter = None;
+                    is_assignment = false;
+                  }
+                in
```

Comment:
These 10 lines are copied all over the place and they depend only on `var`. You should lift them into an auxiliary function.